### ISSUE TITLE
Is there a way to disable gradients at inference time for a neural network?

### ISSUE THREAD

**Issue by arrufat** [davisking/dlib#1873]
Hi, it might be obvious, but I've been surfing around the code and couldn't find a way to do it.
I would like to disable all the gradient values that are stored in the network when I know I won't need them, that is, when performing network calls and not training.

I found this method: `get_output_and_gradient_input_disabled(true)`, which has a really promising name.
However, I wonder how should I set it from a network, if at all possible.

Thanks in advance!

**Comment by davisking** [comment]
Gradients aren't computed unless you are doing a backward pass. So I'm not sure what you are asking. It seems that it's already "disabled during inference".

**Comment by arrufat** [comment]
Thanks for the quick response.
I was thinking about something similar to `torch.no_grad()` in PyTorch, which greatly reduces GPU memory usage. But as you say, that seems to be not happening already, so it's not an issue.

**Comment by arrufat** [comment]
However, I would like to do more tests later:  I want to check if the GPU memory usage increases when running the backward pass. It should, right?

**Comment by davisking** [comment]
That is correct.

**Comment by arrufat** [comment]
Great, thank you so much.
I ran the following test:
- load network to GPU: 203 MiB
- perform a forward call with an image of 224x244: 491 MiB
- perform a backward call: 614 MiB

In fact, the way dlib works makes so much sense.
Thanks again :)
